File: Advent of Code/2015-Day2.erl
-module(day2_2015).
-export([
  chall/1,
  chall2/1,
  open/1,
  trans/1,
  tr/2,
  work/2
  ]).

%% open file and create token

open(Filename) ->
  {ok, Bin} = file:read_file(Filename),
   S = string:lexemes(binary_to_list(Bin), "\r\n"),
   work(S,[]).

%% main part

work([], Acc) -> lists:sum(Acc);
work([H|T], Acc) ->
  Temp = string:split(H, "x", all),
  V = tr(Temp,[]),
  work(T, [chall2(V) | Acc]).

%% Split and convert String to Integer

trans([H|_]) ->
  Temp = string:split(H, "x", all),
  tr(Temp,[]).

tr([], Acc) -> lists:reverse(Acc);
tr([H|T], Acc) when is_list(H) ->
  {Int,_} = string:to_integer(H),
  tr(T, [ Int | Acc]).

%% chall part 1

chall([L,W,H]) ->
  One = 2 * lists:foldl(fun(X, Acc) -> X * Acc end, 1, [L,W]),
  Two = 2 * lists:foldl(fun(X, Acc) -> X * Acc end, 1, [W,H]),
  Three = 2 * lists:foldl(fun(X, Acc) -> X * Acc end, 1, [L,H]),
  Slack = lists:foldl(fun(X, Acc) -> X * Acc end, 1, lists:sublist(lists:sort([L,W,H]),1,2)),
  One + Two + Three + Slack.

%% chall part 2

chall2([L,W,H]) ->
  One = lists:foldl(fun(X, Acc) -> X + X + Acc end, 0, lists:sublist(lists:sort([L,W,H]),1,2)),
  Two = lists:foldl(fun(X, Acc) -> X * Acc end, 1, [L,W,H]),
  One + Two.

File: Advent of Code/2015-Day2.ex
# Day2 - Elixir
# Xophidia


defmodule Day1_2015 do
  def calcul(dimension) do
    [l,w,h] = dimension
    one = [l,w] |> Enum.reduce(fn(x, acc) -> x * acc * 2 end)
    two = [w,h] |> Enum.reduce(fn(x, acc) -> x * acc * 2 end)
    three = [l,h] |> Enum.reduce(fn(x, acc) -> x * acc * 2 end)
    slack = dimension |> Enum.sort |> Enum.take(2) |> Enum.reduce(fn(x, acc) -> x * acc end)
    one + two + three + slack
  end
end

file = File.read!("day2.txt") |> String.strip |> String.split("\n")
tuple = for line <- file, do: Enum.map(String.split(line, "x"), &String.to_integer/1)
IO.inspect Enum.map(tuple, &Day1_2015.calcul/1) |> Enum.sum

IO.inspect Day1_2015.calcul([1,2,3]) == 24
